libs-scala/ports/src/main/scala/com/digitalasset/ports/FreePort.scala

```diff
@@ -3,17 +3,127 @@
 
 package com.daml.ports
 
+import java.io.IOException
 import java.net.{InetAddress, ServerSocket}
 
+import com.daml.bazeltools.BazelRunfiles.rlocation
+
+import scala.io.Source
+import scala.sys.process.Process
+import scala.util.{Failure, Random, Try}
+
 object FreePort {
+  private val maxAttempts = 100
+  private[ports] val dynamicRange = dynamicPortRange()
 
+  /** Find a free port.
+    *
+    * This determines whether a port is free by attempting to bind it and immediately releasing it again.
+    * Note, this does not guarantee that the port is still free after this function returns, see [[LockedFreePort]].
+    *
+    * Draws ports from outside the dynamic port range (port 0)
+    * to avoid collisions with parts of the code base that directly bind to port 0.
+    */
   def find(): Port = {
-    val socket = new ServerSocket(0, 0, InetAddress.getLoopbackAddress)
+    val portGen = randomPortGen(dynamicRange)
+    val portCandidates = (1 to maxAttempts).map(_ => portGen())
+    portCandidates
+      .find { candidate =>
+        try {
+          val socket = new ServerSocket(candidate, 0, InetAddress.getLoopbackAddress)
+          socket.close()
+          true
+        } catch {
+          case _: IOException => false
+        }
+      }
+      .map(found => Port(found))
+      .getOrElse(throw NoFreePortException)
+  }
+
+  sealed abstract class FindPortException(message: String, cause: Throwable)
+      extends RuntimeException(message, cause) {
+    def this(message: String) = this(message, null)
+  }
+
+  final case object NoFreePortException extends FindPortException("No free port found")
+
+  final case class DynamicPortRangeException(cause: Throwable)
+      extends FindPortException("Failed to determine the dynamic port range", cause)
+
+  private def randomPortGen(dynamicRange: (Int, Int)): () => Int = {
+    val (minPort, maxPort) = (1024, 65536)
+    val minExcl = Math.min(Math.max(minPort, dynamicRange._1), maxPort)
+    val maxExcl = Math.min(Math.max(minExcl, dynamicRange._2), maxPort)
+    val numLowerPorts = minExcl - minPort
+    val numUpperPorts = maxPort - maxExcl
+    val numAvailablePorts = numLowerPorts + numUpperPorts
+    val gen = new Random()
+    def genPort(): Int = {
+      val n = gen.nextInt(numAvailablePorts)
+      if (n < numLowerPorts) {
+        n + minPort
+      } else {
+        n - numLowerPorts + maxExcl + 1
+      }
+    }
```

Comment:
```suggestion
    def genPort(): Int = {
      val n = scala.util.Random.nextInt(numAvailablePorts)
      if (n < numLowerPorts) {
        n + minPort
      } else {
        n - numLowerPorts + maxExcl + 1
      }
    }
```
Instantiating a `Random` is useful to improve testability using a fixed seed, otherwise you can just use the companion object which already is a `Random` instance.